t/apicast-policy-buffering.t

```diff
@@ -0,0 +1,164 @@
+    {
+      "backend_version":  1,
+      "proxy": {
+        "api_backend": "http://test:$TEST_NGINX_SERVER_PORT/",
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "GET", "metric_system_name": "hits", "delta": 2 }
+        ],
+        "policy_chain": [
+          {
+            "name": "buffering",
+            "version": "builtin",
+            "configuration": {}
+          },
+          {
+            "name": "apicast",
+            "version": "builtin",
+            "configuration": {}
+          }
+        ]
+      }
+    }
+  ]
+}
+--- backend
+location /transactions/authrep.xml {
+  content_by_lua_block {
+    ngx.exit(200)
+  }
+}
+--- upstream
+location / {
+  proxy_pass http://127.0.0.1:19113;
+}
+--- request
+GET /?user_key=
+--- tcp_listen: 19113
+--- tcp_no_close
+--- tcp_reply eval
+"HTTP/1.1 200 OK\r\nTransfer-Encoding: chunked\r\n\r\n5\r\nhello\r\n0\r\n\r\n"
+--- response_body chomp
+hello
+
+
+=== TEST 2: default configuration (buffered) with big file
+--- configuration
+{
+  "services": [
+    {
+      "backend_version":  1,
+      "proxy": {
+        "api_backend": "http://test-upstream.lvh.me:$TEST_NGINX_SERVER_PORT/",
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "POST", "metric_system_name": "hits", "delta": 2 }
+        ],
+        "policy_chain": [
+          {
+            "name": "buffering",
+            "version": "builtin",
+            "configuration": {}
+          },
+          {
+            "name": "apicast",
+            "version": "builtin",
+            "configuration": {}
+          }
+        ]
+      }
+    }
+  ]
+}
+--- backend
+location /transactions/authrep.xml {
+  content_by_lua_block {
+    ngx.exit(200)
+  }
+}
+--- upstream
+server_name test-upstream.lvh.me;
+  location / {
+    access_by_lua_block {
+      ngx.say("yay, api backend")
+    }
+  }
+--- request eval
+"POST /?user_key= \n" . $ENV{LARGE_BODY}
+--- error_code: 200
+--- error_log env
+a client request body is buffered to a temporary file
+--- no_error_log
+
+
+=== TEST 3: request buffering off (un-buffered) with big file
+--- configuration
+{
+  "services": [
+    {
+      "backend_version":  1,
+      "proxy": {
+        "api_backend": "http://test-upstream.lvh.me:$TEST_NGINX_SERVER_PORT/",
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "POST", "metric_system_name": "hits", "delta": 2 }
+        ],
+        "policy_chain": [
+          {
+            "name": "buffering",
+            "version": "builtin",
+            "configuration": {
+                "request_buffering":  false
+            }
+          },
+          {
+            "name": "apicast",
+            "version": "builtin",
+            "configuration": {}
+          }
+        ]
+      }
+    }
+  ]
+}
+--- backend
+location /transactions/authrep.xml {
+  content_by_lua_block {
+    ngx.exit(200)
+  }
+}
+--- upstream
+server_name test-upstream.lvh.me;
+  location / {
+    access_by_lua_block {
```

Comment:
can we assert here that the request has the TransferEncoding header?

Reply:
The request is non-chunked, I will add another test with chunked body

Reply:
Just bear in mind that if the downstream request has chunked transfer encoding, the upstream request does not need to be chunked transfer encoding as well. The policy only ensures that the upstream request will start as soon as the downstream requests arrives. 

So, if Nginx can read the entire body quickly (it is small) and at the time it reads the headers, the upstream request will not be chunk encoded and instead it will have Content-Length. Only if there is a big file we can *try* to force nginx to start chunked transfer encoding to upstream. Or we add a delay in the request, but not sure we can do that in the Test-Nginx framework